src/red_black_tree.jl

```diff
@@ -0,0 +1,327 @@
+
+function delete_fix(tree::RBTree, node::Union{RBTreeNode, Nothing})
+    while node != tree.root && !node.color 
+         if node == node.parent.leftChild
+            sibling = node.parent.rightChild
+            if sibling.color
+                sibling.color = false
+                node.parent.color = true
+                left_rotate!(tree, node.parent)
+                sibling = node.parent.rightChild 
+            end
+
+            if !sibling.rightChild.color && !sibling.leftChild.color
+                sibling.color = true
+                node = node.parent
+            else
+                if !sibling.rightChild.color
+                    sibling.leftChild.color = false
+                    sibling.color = true
+                    right_rotate!(tree, sibling)
+                    sibling = node.parent.rightChild
+                end
+
+                sibling.color = node.parent.color
+                node.parent.color = false
+                sibling.rightChild.color = false
+                left_rotate!(tree, node.parent)
+                node = tree.root
+            end
+        else
+            sibling = node.parent.leftChild
+            if sibling.color
+                sibling.color = false
+                node.parent.color = true
+                right_rotate!(tree, node.parent)
+                sibling = node.parent.leftChild 
+            end
+
+            if !sibling.rightChild.color && !sibling.leftChild.color
+                sibling.color = true
+                node = node.parent
+            else
+                if !sibling.leftChild.color
+                    sibling.rightChild.color = false
+                    sibling.color = true
+                    left_rotate!(tree, sibling)
+                    sibling = node.parent.leftChild
+                end
+
+                sibling.color = node.parent.color
+                node.parent.color = false
+                sibling.leftChild.color = false
+                right_rotate!(tree, node.parent)
+                node = tree.root
+            end
+        end
+    end
+    node.color = false
+    return tree
+end
+
+# transplant u in the tree by v 
+function rb_transplant(tree::RBTree, u::Union{RBTreeNode, Nothing}, v::Union{RBTreeNode, Nothing})
+    if u.parent == nothing
+        tree.root = v
+    elseif u == u.parent.leftChild
+        u.parent.leftChild = v
+    else
+        u.parent.rightChild = v
+    end
+    v.parent = u.parent
+end
+
+function minimum_node(tree::RBTree, node::RBTreeNode)
+    while node.leftChild != tree.Nil
+        node = node.leftChild
+    end
+    return node
+end
+
+function Base.delete!(tree::RBTree{K}, d::K) where K
+    z = tree.Nil
+    node = tree.root
+
+    while node != tree.Nil
+        if node.data == d
+            z = node
+        end
+
+        if d < node.data
+            node = node.leftChild
+        else
+            node = node.rightChild
+        end
+    end
+
+    (z == tree.Nil) && throw(KeyError(d))
```

Comment:
`delete!` on dicts for items that are not present does not error in julia.
So maybe it should be the same for trees?

```julia
julia> x = Dict(1=>3)
Dict{Int64,Int64} with 1 entry:
  1 => 3

julia> delete!(x, 2)
Dict{Int64,Int64} with 1 entry:
  1 => 3
```

Reply:
It can be done. But let's think of it in this way. This file is mostly to be used as an underlying container for `SortedDict` right? Throwing a `KeyError` in the underlying container doesn't hurt, we can catch this and process accordingly.

Reply:
Consistency is important to make it easier to learn.

Exception catching is expensive.
As you say `SortedDict` would just supress that error and return the `Dict` unmodified anyway.